README.md
# IP Echo

A simple no-logs Vercel Edge function that returns the caller's IP address.

## Usage

A `GET` request to <https://ipv4.average.name/api> will return a string containing the caller's IPv4 address. Note that, because this project is hosted on Vercel, and [Vercel does not support IPv6](https://github.com/orgs/vercel/discussions/47#discussioncomment-4314763), you will only see an IPv4 address from this endpoint.

```sh
$ curl https://ipv4.average.name/api
1.2.3.4
```

Navigating to the root returns an HTML file that defines a redirect to the `/api` endpoint, because Vercel is weird.

The API is documented using [OpenAPI](https://petstore.swagger.io/?url=https://raw.githubusercontent.com/AverageHelper/ip-echo-vercel/main/openapi.yaml). You can generate a TypeScript client by plugging our spec file into [oazapfts](https://www.npmjs.com/package/oazapfts):

```sh
$ oazapfts https://raw.githubusercontent.com/AverageHelper/ip-echo-vercel/main/openapi.yaml ./ip.ts
```

```ts
import { ip } from "./ip";

const result = await ip();

switch (result.status) {
	case 200:
		console.info(result.data); // "1.2.3.4"
		break;

	case 404:
		console.warn("Couldn't get IP address");
		break;

	default:
		console.error("Something went very very wrong!");
}
```

## No logs?

No more than Vercel's normal logging, which looks like this:

![A screenshot of Vercel's connection logs. Included is the time of the request, the request status code, the requested host name, the requested document path on the host, the requester's user agent string, and Vercel's request ID string.](/docs/images/example-logs.png)

There's not much there to identify you, unless you put some personally-identifying information into your request path, or contact me directly and inform me that you made a request.

I'm not doing any additional logging myself. You can read as much [in the code](/api/index.ts), though I suppose I cannot prove that I'm running _this_ repo on the above-named endpoint on Vercel's Edge network. If you're concerned I may be snooping, feel free to run your own instance!


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.6] - 2023-12-14

### Changed

- We've moved to Forgejo. Replaced relevant GitHub references with the appropriate URL.

### Security

- Updated vulnerable dependencies.

## [2.0.5] - 2023-08-12

### Changed

- Updated engine to Node 18.

## [2.0.4] - 2023-08-06

### Security

- Send `Content-Security-Policy`, `X-Frame-Options`, `X-Content-Type-Options`, `Referrer-Policy`, and `Permissions-Policy` headers.

## [2.0.3] - 2023-03-11

### Changed

- `Content-Type` is now `text/plain` by default.

## [2.0.2] - 2023-03-05

### Fixed

- All responses include CORS headers again.

## [2.0.1] - 2023-03-05

### Added

- Responses now include the `Access-Control-Allow-Methods` header, with a value of `GET, OPTIONS`.

### Removed

- Responses no longer include `X-CSRF-Token`,`X-Requested-With`, `Accept-Version`, `Content-MD5`, or `X-Api-Version` in the `Access-Control-Allow-Headers` header.
- Responses no longer include CORS headers unless the request method is `OPTIONS` (ostensibly a CORS preflight request).

## [2.0.0] - 2023-03-05

### Changed

- Now sends a plain string in `text/html` by default. To receive JSON, include `application/json` in the `Accept` header.
- Now sends a newline at the end of the output.

## [1.0.2] - 2023-03-04

### Added

- An email address for people to reach me regarding our [Code of Conduct](/CODE_OF_CONDUCT.md).

### Fixed

- Attempt to make the auto-test pipeline not presume v1.x.x versions are "prerelease".

## [1.0.1] - 2023-03-04

### Added

- Auto-test pipeline.

## [1.0.0] - 2023-03-04

### Added

- Vercel Edge function to echo the caller's IP address.

[2.0.6]: https://git.average.name/AverageHelper/ip-echo-vercel/compare/v2.0.5...v2.0.6
[2.0.5]: https://git.average.name/AverageHelper/ip-echo-vercel/compare/v2.0.4...v2.0.5
[2.0.4]: https://git.average.name/AverageHelper/ip-echo-vercel/compare/v2.0.3...v2.0.4
[2.0.3]: https://git.average.name/AverageHelper/ip-echo-vercel/compare/v2.0.2...v2.0.3
[2.0.2]: https://git.average.name/AverageHelper/ip-echo-vercel/compare/v2.0.1...v2.0.2
[2.0.1]: https://git.average.name/AverageHelper/ip-echo-vercel/compare/v2.0.0...v2.0.1
[2.0.0]: https://git.average.name/AverageHelper/ip-echo-vercel/compare/v1.0.2...v2.0.0
[1.0.2]: https://git.average.name/AverageHelper/ip-echo-vercel/compare/v1.0.1...v1.0.2
[1.0.1]: https://git.average.name/AverageHelper/ip-echo-vercel/compare/v1.0.0...v1.0.1
[1.0.0]: https://git.average.name/AverageHelper/ip-echo-vercel/releases/tag/v1.0.0


src/index.test.ts
import type { VercelRequest, VercelResponse } from "@vercel/node";
import "jest-extended";
import { echo } from "./index";
import express from "express";
import request from "supertest";

describe("IP Echo", () => {
	const TEST_IP = "::ffff:127.0.0.1";
	const app = express() //
		.use("/", echo);

	test("has a default export that matches the named export", async () => {
		// Vercel requires a default export
		const unit = await import("./index");
		expect(unit).toHaveProperty("default", echo);
	});

	test("has a Vercel Edge config", async () => {
		const unit = await import("./index");
		expect(unit).toHaveProperty("config");
		expect(unit.config).toStrictEqual({
			runtime: "edge",
		});
	});

	test("responds correctly to CORS preflight", async () => {
		const response = await request(app)
			.options("/")
			.expect("Vary", "*")
			.expect("Cache-Control", "no-store")
			.expect("Access-Control-Allow-Origin", "*")
			.expect(200);
		expect(response.body).toStrictEqual({});
	});

	// See https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods
	const BadMethods = ["HEAD", "POST", "PUT", "DELETE", "TRACE", "PATCH"] as const;
	test.each(BadMethods)("responds 405 to %s requests", async rawMethod => {
		const method = rawMethod.toLowerCase() as Lowercase<typeof rawMethod>;

		const response = await request(app)
			[method]("/")
			.expect("Vary", "*")
			.expect("Cache-Control", "no-store")
			.expect(405);
		expect(response.body).toStrictEqual({});
	});

	test("responds with the caller's IP address", async () => {
		const response = await request(app)
			.get("/")
			.expect("Vary", "*")
			.expect("Cache-Control", "no-store")
			.expect("Content-Type", /json/u)
			.expect(200);
		expect(response.body).toBe(TEST_IP);
	});

	describe("when in Vercel", () => {
		const IP_HEADER_NAME = "x-real-ip";
		let mockReq: VercelRequest;
		let mockRes: VercelResponse;

		beforeEach(() => {
			mockReq = {
				method: "GET",
				headers: {},
			} as unknown as VercelRequest;
			mockRes = {
				setHeader: jest.fn().mockImplementation(() => mockRes),
				status: jest.fn().mockImplementation(() => mockRes),
				json: jest.fn().mockImplementation(() => mockRes),
				end: jest.fn().mockImplementation(() => mockRes),
			} as unknown as VercelResponse;
		});

		test("responds with 'unknown' if there is no 'X-Real-IP' header", () => {
			expect(echo(mockReq, mockRes)).toBeUndefined(); // returns void
			expect(mockRes.status).toHaveBeenCalledOnceWith(404);
			expect(mockRes.json).not.toHaveBeenCalled();
		});

		test("responds with the IP address defined in the 'X-Real-IP' header", () => {
			mockReq.headers[IP_HEADER_NAME] = TEST_IP;
			expect(echo(mockReq, mockRes)).toBeUndefined(); // returns void
			expect(mockRes.status).toHaveBeenCalledOnceWith(200);
			expect(mockRes.json).toHaveBeenCalledOnceWith(TEST_IP);
		});
	});
});


tests/index.test.ts
import "jest-extended";
import endpoint from "../api/index";
import fetchMock from "jest-fetch-mock";
fetchMock.enableMocks(); // Enables use of `Request` and `Response` objects

function assertHasSecurityHeaders(res: Response): void {
	expect(res).toHaveProperty("headers");

	// ** Security **
	// Vercel seems to set Strict-Transport-Security automatically
	expect(res.headers.get("Content-Security-Policy")).toBe("default-src 'self'");
	expect(res.headers.get("X-Content-Type-Options")).toBe("nosniff");
	expect(res.headers.get("X-Frame-Options")).toBe("SAMEORIGIN");
	expect(res.headers.get("Referrer-Policy")).toBe("no-referrer");
	expect(res.headers.get("Permissions-Policy")).toBe(
		"accelerometer=(), ambient-light-sensor=(), autoplay=(), battery=(), camera=(), clipboard-read=(), clipboard-write=(), cross-origin-isolated=(), display-capture=(), document-domain=(), encrypted-media=(), execution-while-not-rendered=(), execution-while-out-of-viewport=(), fullscreen=*, gamepad=(), geolocation=(), gyroscope=(), identity-credentials-get=(), idle-detection=(), interest-cohort=(), keyboard-map=(), local-fonts=(), magnetometer=(), microphone=(), midi=(), navigation-override=(), payment=(), picture-in-picture=*, publickey-credentials-create=(), publickey-credentials-get=(), screen-wake-lock=(), serial=(), speaker-selection=(), storage-access=(), sync-xhr=(), usb=(), web-share=(), xr-spatial-tracking=()"
	);

	// ** Miscellaneous **
	expect(res.headers.get("Vary")).toBe("*");
	expect(res.headers.get("Cache-Control")).toBe("no-store");
	expect(res.headers.get("Access-Control-Allow-Origin")).toBe("*");
	expect(res.headers.get("Access-Control-Allow-Methods")).toInclude("GET");
	expect(res.headers.get("Access-Control-Allow-Headers")).toInclude("Accept");
}

describe("IP Echo", () => {
	const url = new URL("https://localhost/");
	const IP_HEADER_NAME = "x-real-ip";
	const TEST_IP = "::ffff:127.0.0.1";

	test("has a Vercel Edge config", async () => {
		const unit = await import("../api/index");
		expect(unit).toHaveProperty("config");
		expect(unit.config).toStrictEqual({
			runtime: "edge",
		});
	});

	test("responds correctly to CORS preflight", () => {
		const req = new window.Request(url, { method: "OPTIONS" });
		const res = endpoint(req);
		expect(res.status).toBe(204);
		assertHasSecurityHeaders(res);
		expect(res.body).toBe(null);
	});

	// See https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods
	const BadMethods = ["HEAD", "POST", "PUT", "DELETE", "TRACE", "PATCH"] as const;
	test.each(BadMethods)("responds 405 to %s requests", method => {
		const req = new window.Request(url, { method });
		const res = endpoint(req);
		expect(res.status).toBe(405);
		assertHasSecurityHeaders(res);
		expect(res.body).toBe(null);
	});

	test("responds with the caller's IP address", async () => {
		const req = new window.Request(url, { headers: { [IP_HEADER_NAME]: TEST_IP } });
		const res = endpoint(req);
		expect(res.headers.get("Content-Type")).toBe("text/plain");
		expect(res.status).toBe(200);
		assertHasSecurityHeaders(res);
		expect(await res.text()).toBe(TEST_IP.concat("\n"));
	});

	test("responds with the caller's IP address in JSON", async () => {
		const req = new window.Request(url, {
			headers: {
				[IP_HEADER_NAME]: TEST_IP,
				Accept: "application/json",
			},
		});
		const res = endpoint(req);
		expect(res.status).toBe(200);
		assertHasSecurityHeaders(res);
		expect(res.headers.get("Content-Type")).toBe("application/json");
		expect(await res.json()).toBe(TEST_IP);
	});

	test("responds with 404 if there is no 'X-Real-IP' header", () => {
		const req = new window.Request(url);
		const res = endpoint(req);
		expect(res.status).toBe(404);
		assertHasSecurityHeaders(res);
		expect(res.body).toBe(null);
	});
});


scripts/release.ts
#!/usr/bin/env tsx

import { assert, enums, string, type } from "superstruct";
import { parser as changelogParser } from "keep-a-changelog";
import { readFileSync, writeFileSync } from "node:fs";
import { URL } from "node:url";
import * as semver from "semver";

const logger = console;

// Fixes the changelog's footer links and bumps the `version` in [package.json](/package.json) and [package-lock.json](/package-lock.json).
// This script may be run repeatedly on the same project.

const { parse: parseSemVer } = semver;

function quote(str: string | undefined): string | undefined {
	if (str === undefined) return str;
	return `'${str}'`;
}

logger.info("** release.ts **");

// Load the changelog
const changelogPath = new URL("../CHANGELOG.md", import.meta.url).pathname;
const packageJsonPath = new URL("../package.json", import.meta.url).pathname;
const packageLockJsonPath = new URL("../package-lock.json", import.meta.url).pathname;
logger.info("Loading changelog from", quote(changelogPath));

const rawChangelog = readFileSync(changelogPath, "utf-8");
const changelog = changelogParser(rawChangelog);

const releases = changelog.releases;

// Get current versioned release
const thisReleaseIdx = releases.findIndex(release => release.date && release.parsedVersion);
const thisRelease = releases[thisReleaseIdx];
if (!thisRelease?.parsedVersion || !thisRelease.version)
	throw new TypeError("No versioned release was found.");

// Handy info
logger.info("latest release:", thisRelease.version);

const prevRelease = releases[thisReleaseIdx + 1];
logger.info("previous release:", prevRelease?.version);

// Fix the changelog's format (new compare links, etc.), and print the diff of our changes
logger.info("\n** Spec compliance **");

changelog.format = "markdownlint";
const newChangelog = changelog.toString();
writeFileSync(changelogPath, newChangelog);

const didFixChangelog = rawChangelog !== newChangelog;
if (!didFixChangelog) {
	logger.info("Changelog was already spec compliant.");
} else {
	logger.info("Fixed formatting for spec compliance.");
}

// Fix package.json and package-lock.json
logger.info("\n** Version matching **");
const versioned = type({
	version: string(),
});
const versionedLock = type({
	version: string(),
	lockfileVersion: enums([2, 3]),
	packages: type({
		"": type({
			version: string(),
		}),
	}),
});

const packageJson: unknown = JSON.parse(readFileSync(packageJsonPath, "utf-8"));
const packageLockJson: unknown = JSON.parse(readFileSync(packageLockJsonPath, "utf-8"));

assert(packageJson, versioned);
assert(packageLockJson, versionedLock);

const packageVersion = parseSemVer(packageJson.version);
const packageLockVersion = parseSemVer(packageLockJson.version);

if (!packageVersion)
	throw new TypeError(
		'The "version" field in package.json is not a compliant Semantic Version number'
	);

if (!packageLockVersion)
	throw new TypeError(
		'The "version" field in package-lock.json is not a compliant Semantic Version number'
	);

logger.info("package.json version:", packageVersion.version);
logger.info("package-lock.json version:", packageLockVersion.version);

if (packageVersion.version !== packageLockVersion.version)
	throw new EvalError(
		'The "version" fields in package.json and package-lock.json do not match. Please let CHANGELOG.md be the source of truth for versioning. To ignore this warning and proceed, please first run `npm install`.'
	);

// Update package.json
const oldPackageJson = `${JSON.stringify(packageJson, undefined, "\t")}\n`;
packageJson.version = thisRelease.version;
const newPackageJson = `${JSON.stringify(packageJson, undefined, "\t")}\n`;
writeFileSync(packageJsonPath, newPackageJson);

const didFixPackageJson = oldPackageJson !== newPackageJson;
if (!didFixPackageJson) {
	logger.info("package.json already had the correct version.");
} else {
	logger.info("Updated package.json version.");
}

// Update package-lock.json
const oldPackageLockJson = `${JSON.stringify(packageLockJson, undefined, "\t")}\n`;
// Maybe we should just run `npm i` instead?
packageLockJson.version = thisRelease.version;
packageLockJson.packages[""].version = thisRelease.version;
const newPackageLockJson = `${JSON.stringify(packageLockJson, undefined, "\t")}\n`;
writeFileSync(packageLockJsonPath, newPackageLockJson);

const didFixPackageLockJson = oldPackageLockJson !== newPackageLockJson;
if (!didFixPackageLockJson) {
	logger.info("package-lock.json already had the correct version.");
} else {
	logger.info("Updated package-lock.json version.");
}

// If we fixed the changelog or updated package.json, throw
if (didFixChangelog || didFixPackageJson || didFixPackageLockJson) {
	logger.warn("⚠️  We made some changes. Please review them and re-run. ⚠️");
	process.exit(1); // this should fail us in CI
}


scripts/createRelease.ts
#!/usr/bin/env tsx

import { parser as changelogParser } from "keep-a-changelog";
import { readFileSync } from "node:fs";

// Parses the changelog for 'version', 'description', 'versionMajor', and 'status' bits, and
// creates a Release on the repo host if the release doesn't already exist.
// This script may be run repeatedly on the same project.

const logger = console;

function quote(str: string | undefined): string | undefined {
	if (str === undefined) return str;
	return `'${str}'`;
}

function requireEnv(key: string): string {
	const result = process.env[key];
	if (!result) throw new TypeError(`Missing env var '${key}'`);
	return result;
}

logger.info("** createRelease.ts **");

// Load env vars
const accessToken = requireEnv("FORGEJO_ACCESS_TOKEN");
const ciRepo = requireEnv("GITHUB_REPOSITORY"); // repository full name (`<owner>/<name>`)
const repoReleases = `https://git.average.name/api/v1/repos/${ciRepo}/releases`;
logger.info(`Releases API: ${repoReleases}`);

// Load the changelog
const changelogPath = new URL("../CHANGELOG.md", import.meta.url).pathname;
logger.info("Loading changelog from", quote(changelogPath), "\n");

const rawChangelog = readFileSync(changelogPath, "utf-8");
const changelog = changelogParser(rawChangelog);

const releases = changelog.releases;

// Get current versioned release
const thisReleaseIdx = releases.findIndex(release => release.date && release.parsedVersion);
const thisRelease = releases[thisReleaseIdx];
if (!thisRelease?.parsedVersion || !thisRelease.version)
	throw new TypeError("No versioned release was found.");

// Handy info
const tag: string = `v${thisRelease.version}`;
logger.info("latest version:", tag);
logger.info("version metadata:", thisRelease);

// See https://git.average.name/api/swagger#/repository/repoCreateRelease
// GET /repos/{owner}/{repo}/releases/tags/{tag}
const remoteReleaseRes = await fetch(`${repoReleases}/tags/${tag}`);
const remoteRelease: unknown = await remoteReleaseRes.json();

if (remoteReleaseRes.status !== 404) {
	// Release exists! Stop here.
	logger.info("Release already exists on remote:", remoteRelease);
	process.exit(0);
}

// Release doesn't exist! Create it.
logger.info("Release doesn't yet exist on remote. Creating it...");

const isPrerelease: boolean =
	thisRelease.parsedVersion.prerelease.length > 0 || thisRelease.parsedVersion.major === 0;

function toTitleCase(str: string): string {
	const first = str[0];
	if (!first) return str;

	const rest = str.slice(1);
	return `${first.toUpperCase()}${rest}`;
}

// Because keep-a-changelog's description parser is bugged:
let description: string = "";
for (const [heading, changes] of thisRelease.changes) {
	if (changes.length === 0) continue;

	// Heading ('Added', 'Changed', etc.)
	description += "### ";
	description += toTitleCase(heading);
	description += "\n\n";

	// Contents (usually a list)
	for (const change of changes) {
		description += "- ";

		const lines = change.title.split("\n");
		for (const line of lines) {
			if (line.startsWith("- ")) {
				// Sub-list item
				description += "  ";
			}
			description += line;
			description += "\n";
		}
	}

	description += "\n";
}

const output = `current version: ${tag}
is prerelease: ${isPrerelease ? "true" : "false"}
description: ${description.trim()}
`;
logger.info(output);

// See https://git.average.name/api/swagger#/repository/repoCreateRelease
// POST /repos/{owner}/{repo}/releases
const result = await fetch(repoReleases, {
	method: "POST",
	headers: {
		accept: "application/json",
		Authorization: `token ${accessToken}`,
		"Content-Type": "application/json",
	},
	body: JSON.stringify({
		body: description.trim(),
		draft: false,
		name: tag,
		prerelease: isPrerelease,
		tag_name: tag,
		target_commitish: "main",
	}),
});

if (!result.ok) {
	const resultMessage = await result.text();
	throw new Error(resultMessage);
}

const resultMessage: unknown = await result.json();
logger.info("Release created:", resultMessage);


api/index.ts
import { ipAddress } from "@vercel/edge";

function echo(req: Request): Response {
	const ip = ipAddress(req);
	if (!ip) {
		// No IP address found in headers (or wherever Vercel Edge likes to put those)
		return new Response(undefined, { status: 404, headers });
	}

	// Found IP, respond the way the caller wants us to:
	const accept = req.headers.get("accept");
	let message = ip;
	let contentType = "text/plain";

	if (accept?.includes("application/json")) {
		message = JSON.stringify(ip);
		contentType = "application/json";
	}

	return new Response(message.concat("\n"), {
		status: 200,
		headers: { ...headers, "Content-Type": contentType },
	});
}

// Common headers:
const headers = {
	Vary: "*", // See https://stackoverflow.com/a/54337073 for why "Vary: *" is necessary for Safari
	"Cache-Control": "no-store",
	"X-Clacks-Overhead": "GNU Terry Pratchett",
	"Access-Control-Allow-Methods": "GET, OPTIONS", // TODO: Handle HEAD as well
	"Access-Control-Allow-Headers": "Accept, Content-Length, Content-Type, Date",
	"Access-Control-Allow-Origin": "*",
	"Content-Security-Policy": "default-src 'self'",
	"X-Content-Type-Options": "nosniff",
	"X-Frame-Options": "SAMEORIGIN",
	"Referrer-Policy": "no-referrer",
	"Permissions-Policy":
		"accelerometer=(), ambient-light-sensor=(), autoplay=(), battery=(), camera=(), clipboard-read=(), clipboard-write=(), cross-origin-isolated=(), display-capture=(), document-domain=(), encrypted-media=(), execution-while-not-rendered=(), execution-while-out-of-viewport=(), fullscreen=*, gamepad=(), geolocation=(), gyroscope=(), identity-credentials-get=(), idle-detection=(), interest-cohort=(), keyboard-map=(), local-fonts=(), magnetometer=(), microphone=(), midi=(), navigation-override=(), payment=(), picture-in-picture=*, publickey-credentials-create=(), publickey-credentials-get=(), screen-wake-lock=(), serial=(), speaker-selection=(), storage-access=(), sync-xhr=(), usb=(), web-share=(), xr-spatial-tracking=()",
};

// Vercel wants the endpoint to be a `default` export:
export default function handler(req: Request): Response {
	switch (req.method.toUpperCase()) {
		// Normal requests:
		case "GET":
			return echo(req);

		// CORS preflight:
		case "OPTIONS":
			return new Response(undefined, { status: 204, headers });

		// Everything else:
		default:
			return new Response(undefined, { status: 405, headers });
	}
}

// Vercel Edge config:
export const config = {
	runtime: "edge",
};
